### DISCUSSION TITLE
Scalar bar from matplotlib cmap

### DISCUSSION THREAD

**Discussion by Galedon1** [pyvista/pyvista#4377] [Q&A] [answered]
Hello, 

I am trying to create a scalar bar from a specific matplotlib cmap ( cmap = mpl.colormaps["coolwarm"] ). 

It is supposed to be done by plotter.add_scalar_bar(), but I have failed to see where the cmap parameter comes in. 

What am I missing?

**Comment by banesullivan** [comment]
`camp` argument is usually used in the `add_mesh()` call. It can handle any named MPL colormap or camp object.

Take a look at our example gallery for countless examples. One of note might be: https://docs.pyvista.org/examples/02-plot/scalar-bars.html#sphx-glr-examples-02-plot-scalar-bars-py

**Reply by Galedon1** [reply]
Thank you for the answer but I do not have any mesh that could be used for that. I need a way to create a scalar bar from scratch, like from numpy vector, or something that will never appear in the plot. 
Is there any workaround?

**Reply by banesullivan** [reply]
I'm very confused about what you're trying to accomplish.

**Reply by Galedon1** [reply]
I have a number of lines (several thousands), each has exactly one color (determined by some property). Now I need to create a colorbar that would explain the meaning of that colors. 
I have the colormap and the minimal and maximum value - that should be enough for the colorbar contruction, right?

**Reply by banesullivan** [reply]
`add_scalar_bar()` requires a mapper so some dataset must have been plotted beforehand.

Maybe this examples send you in the right direction:

```py
import pyvista as pv
import numpy as np

n = 100
vmin, vmax = 33.3, 97.7

pl = pv.Plotter()

for i in range(n):
    line = pv.Line(np.random.rand(1, 3), np.random.rand(1, 3), )
    line.cell_data['foo'] = [np.random.uniform(low=vmin, high=vmax),]
    pl.add_mesh(line, scalars='foo', clim=[vmin, vmax], cmap='coolwarm')

pl.show()
```
![screenshot](https://user-images.githubusercontent.com/22067021/236016271-d7f4da94-b913-4ca9-84a5-f7f6601e1505.png)

**Reply by Galedon1** [reply, marked as answer]
I see, so there has to be a "dummy" mesh that is used for the colormap, Thanks a lot :)

The workaround that worked for me was: 
```python
import pyvista as pv

cmap = pv.Line((0,0,0), (0,0,0)) 
cmap.cell_data["scalar"] = 1
pl.add_mesh(cmap, 
            scalars = "scalar", 
            clim=[0, MaxEcolor], 
            cmap='coolwarm',
            # opacity = 1
            )
```
